Script Editor UX for Level Builder

joshlory opened the pull request:
Replace the DSL text area with a UI for editing Stage and Levels in a Course Script.

![screen shot 2016-11-09 at 3 41 47 pm](https://cloud.githubusercontent.com/assets/413693/20159032/13649620-a693-11e6-8a33-fa2798b9e6cf.png)

Bjvanminnen commented on apps/src/sites/studio/pages/components/FlexGroup.jsx:
```diff
@@ -0,0 +1,113 @@
+import React from 'react';
+import { connect } from 'react-redux';
+import _ from 'lodash';
+import color from '../../../../util/color';
```
I wonder if we should start using `@cdo/apps/utils/color` instead of these long paths of `../` that are hard to follow. I think @pcardune had suggested that a good heuristic might be to provide relative paths for things below you (i.e. if your path starts with `./` but use `@cdo/apps` otherwise. Should we formalize that? (Not sure if there are downsides to using @cdo/apps in mroe places.

pcardune replied:
The main downside to using `@cdo/apps` in more places is that static analysis and refactoring tools are less likely to understand what they mean (as that would require them to understand webpack module aliases)

Bjvanminnen commented on apps/src/sites/studio/pages/components/FlexGroup.jsx:
```diff
@@ -0,0 +1,113 @@
+
+  handleAddGroup() {
+    this.props.addGroup();
+  },
```
I think you could just do `onMouseDown={this.props.addGroup}` instead of creating a wrapper method.

Bjvanminnen commented on apps/src/sites/studio/pages/components/FlexGroup.jsx:
```diff
@@ -0,0 +1,113 @@
+                    />
+                  );
+                })}
+                <button onMouseDown={this.handleAddStage.bind(null, afterStage - 1)} className="btn" style={styles.addStage} type="button">
```
styleguide prefers one prop per line when we have multiple props

Bjvanminnen replied:
What does "type=button" accomplish on a button?

joshlory replied:
Prevents the button from submitting the form: http://stackoverflow.com/a/10836076.

Bjvanminnen commented on apps/src/sites/studio/pages/components/FlexGroup.jsx:
```diff
@@ -0,0 +1,113 @@
+                  );
+                })}
+                <button onMouseDown={this.handleAddStage.bind(null, afterStage - 1)} className="btn" style={styles.addStage} type="button">
+                  <i style={{marginRight: 7}} className="fa fa-plus-circle" />
```
Could use our FontAwesome component here and elsewhere

Bjvanminnen commented on apps/src/sites/studio/pages/components/FlexGroup.jsx:
```diff
@@ -0,0 +1,113 @@
+  stages: state
+}), dispatch => ({
+  addGroup() {
+    dispatch({type: 'ADD_GROUP'});
```
The pattern we've been following elsewhere is that instead of dispatching a type with a string here, we have an action creator and a const in our redux file, i.e.

```
// editorRedux.js
const ADD_GROUP = 'editor/ADD_GROUP';
export function addGroup() {
  return { type: ADD_GROUP };
}

export function reducer(state=initialState, action) {
  if (action.type === ADD_GROUP) {
    return nextState;
  }
} 
```

pcardune commented:
The `src/sites/studio/pages/components` directory should not exist. Any generic react components you create should live in `src/templates/`. Non-generic react components (or really any code that is super specific to script editor should live in `src/lib/script-editor`. (i know there isn't a lib directory in `src/` yet, but I will be adding one and moving a bunch of stuff there in the not-to-distant future.

Bjvanminnen commented on apps/src/sites/studio/pages/components/LevelToken.jsx:
```diff
@@ -0,0 +1,131 @@
+
+  render() {
+    return (
+      <Motion
```
There's enough complexity going on here that it might be worth a short comment.

Bjvanminnen commented on apps/src/sites/studio/pages/components/LevelToken.jsx:
```diff
@@ -0,0 +1,131 @@
+  }
+};
+
+const LevelToken = React.createClass({
```
What is a LevelToken? I'm not always great about this either, but I think would be good for us to have class level comments for most of our components describing what they are.

Bjvanminnen commented on apps/src/sites/studio/pages/components/LevelToken.jsx:
```diff
@@ -0,0 +1,131 @@
+              zIndex: this.props.drag ? 1000 : 500 - this.props.level.position
+            })}
+          >
+            <div style={styles.reorder} onMouseDown={this.props.handleDragStart.bind(null, this.props.level.position)}>
```
This might be a place where it makes sense to have a class level method (similar to toggleExpand) such that we avoid rebinding on every render.

Bjvanminnen commented on apps/src/sites/studio/pages/components/LevelTokenDetails.jsx:
```diff
@@ -0,0 +1,153 @@
+import VirtualizedSelect from 'react-virtualized-select';
+import 'react-virtualized/styles.css';
+import 'react-select/dist/react-select.css';
+import 'react-virtualized-select/styles.css';
```
Is this the first place we're importing css? I'm a little surprised our webpack even knows how to do this.

joshlory replied:
I followed the pattern in https://github.com/code-dot-org/code-dot-org/blob/d344587164a1671898b827852c72e0e892918d79/apps/src/publicKeyCryptography/IntegerDropdown.jsx#L6-L8.

Is there a better way to do this?

Bjvanminnen replied:
Nope. Didn't realize we had this elsewhere already. This should be handled by the following line in our webpack config file. https://github.com/code-dot-org/code-dot-org/blob/staging/apps/webpack.js#L26

Bjvanminnen commented on apps/src/sites/studio/pages/components/editorRedux.js:
```diff
@@ -0,0 +1,127 @@
+/* global levelKeyList */
```
Can we avoid having this as a global? First thought (without looking at the code to much) is to have an initial action that populates the store with it.

Bjvanminnen commented on apps/src/sites/studio/pages/components/LevelTokenDetails.jsx:
```diff
@@ -0,0 +1,153 @@
+import 'react-virtualized-select/styles.css';
+import _ from 'lodash';
+
+const levelKeyOptions = _.map(levelKeyList, (label, value) => ({label, value: +value}));
```
Feels strange to have this so far away from the one place it's used. I imagine that the argument for not doing this directly in the render is that we avoid repeated work, which might make sense.

Bjvanminnen commented on apps/src/sites/studio/pages/components/LevelTokenDetails.jsx:
```diff
@@ -0,0 +1,153 @@
+            <span style={Object.assign({display: 'inline-block'}, styles.levelFieldLabel)}>Video key</span>
+            <input defaultValue={this.props.level.videoKey} type="text" style={styles.textInput} />
+          </div>
+        </div>
```
I would indent the continuation (i.e. all the jsx) an extra two spaces.

Bjvanminnen commented on apps/src/sites/studio/pages/components/OrderControls.jsx:
```diff
@@ -0,0 +1,69 @@
+
+export default connect(state => ({}), dispatch => ({
+  move(type, position, direction) {
+    dispatch({type: `MOVE_${type}`, position, direction});
```
In a world where we have action creators for each of our actions, this will presumably look something more like

```
if (type === 'GROUP') {
  dispatch(removeGroup(position));
} else if {
  //etc
```

Bjvanminnen commented on apps/src/sites/studio/pages/components/ScriptEditor.jsx:
```diff
@@ -0,0 +1,126 @@
+ */
+const ScriptEditor = React.createClass({
+  propTypes: {
+    scriptData: React.PropTypes.object.isRequired,
```
I find that when passing in objects of data like this, it can be hard to follow what exactly is expected.

I wonder if it would be clearer to pass in each of the individual fields as individual props. (i.e. instead of scriptData, we have hidden, loginRequired, hideable_stages, etc). At the very least, it would probably be worth (a) describing the shape of this object and/or (b) destructing it in the beginning of your render, i.e. `const { hidden, loginRequired, hideable_stages } = this.props.stageData;`

joshlory replied:
You're right, I'll split this up into individual props.

Bjvanminnen commented on apps/src/sites/studio/pages/components/StageCard.jsx:
```diff
@@ -0,0 +1,166 @@
+  },
+
+  metrics(token) {
+    return ReactDOM.findDOMNode(this.refs[`levelToken${token}`]).getBoundingClientRect();
```
I think in latest versions of React we may not need ReactDOM.findDOMNode (this.refs.foo already gets us the dom node)

pcardune replied:
Using string based refs is considered legacy, although it is not officially deprecated. See https://facebook.github.io/react/docs/refs-and-the-dom.html#the-ref-callback-attribute for an explanation of how you are "supposed" to do it.

Bjvanminnen replied:
FWIW, we use these in a lot of places atm. I personally have yet to see a compelling argument as to why using a callback is better. In almost all cases, it seems much more awkward to me.

pcardune replied:
@Bjvanminnen Have you read through http://stackoverflow.com/questions/37468913/why-ref-string-is-legacy? A lot of stuff there (like static analysis type checkers) do not apply to us at the moment, but the way this breaks potential VM optimizations might be interesting for us depending on the specific scenario.

Anyway, I don't have strong opinions about string vs callback refs in our code base at large, though whenever I see string concatenation involved in object property lookups, I raise my eyebrows ever so slightly.

Bjvanminnen replied:
I'm inclined to agree with you on your point about seeing string concatenation involved in object property lookups. Reading through that SO link now.

Bjvanminnen commented on apps/src/sites/studio/pages/components/StageCard.jsx:
```diff
@@ -0,0 +1,166 @@
+  },
+
+  getInitialState() {
+    return {currentPositions: []};
```
This might be personal preference, but I like to provide default values for any state I might set in getInitialState so that there's one place I can go to look and see what state this component uses

Bjvanminnen commented on apps/src/sites/studio/pages/components/StageCard.jsx:
```diff
@@ -0,0 +1,166 @@
+
+  handleLockableChanged() {
+    const state = this.refs.lockable.checked ? 'lockable' : 'not lockable';
+    console.log(`stage ${this.props.stage.position} marked as ${state}`);
```
Why are we just console.logging here?

Bjvanminnen replied:
Also, I suspect that the onChange callback gives us an arg from which we can figure out the target instead of using refs.
i.e.

```
handleLockableChanged(event) {
  const state = event.target.checked ? 'lockable' : ' not lockable';
}
```

Bjvanminnen commented on apps/src/sites/studio/pages/components/StageCard.jsx:
```diff
@@ -0,0 +1,166 @@
+        <div style={styles.stageCardHeader}>
+          Stage {this.props.stage.position}: {this.props.stage.name}
+          <OrderControls
+            type="STAGE"
```
Could use an enum via. makeEnum in utils for type instead of using a string, i.e. `type={ControlsType.Stage}`

Bjvanminnen commented on apps/src/sites/studio/pages/components/editorRedux.js:
```diff
@@ -0,0 +1,127 @@
+import _ from 'lodash';
+
+function updatePositions(node) {
+  for (var i = 0; i < node.length; i++) {
```
could be let

Bjvanminnen commented on apps/src/sites/studio/pages/components/editorRedux.js:
```diff
@@ -0,0 +1,127 @@
+}
+
+let newStageId = -1;
+let newLevelId = -1;
```
Your reducers are no longer pure, which is an important part of redux. Can you some how derive these from existing state instead?

Bjvanminnen commented on apps/src/sites/studio/pages/components/editorRedux.js:
```diff
@@ -0,0 +1,127 @@
+let newLevelId = -1;
+
+export default function reducer(state, action) {
+  let newState = _.cloneDeep(state);
```
This reducer is going to clone your state and return a new state for every action it sees. In your particular case, that might be fine, but if this reducer was part of a larger store, there would be lots of actions for which this reducer isn't changing anything.

A better practice would be to have each individual case statement return a newState, and if we don't hit any of the case statements just return the unchanged `state`

Bjvanminnen replied:
The fact that most of your actions require us to clone the entire state, and then imperatively make updates also sets off red flags. Ideally, each action should only be mutating a portion of the state.

I wonder if you can construct your state and/or actions differently so that we are mutating less

pcardune replied:
I would go one step further and suggest using `immutable-js` data structures for your state. It makes it a lot easier to efficiently create new state objects with slightly different data.

Bjvanminnen commented on apps/src/sites/studio/pages/components/editorRedux.js:
```diff
@@ -0,0 +1,127 @@
+
+import _ from 'lodash';
+
+function updatePositions(node) {
```
Should this be `incrementPositions` instead?

Bjvanminnen commented on apps/src/sites/studio/pages/components/editorRedux.js:
```diff
@@ -0,0 +1,127 @@
+let newStageId = -1;
+let newLevelId = -1;
+
+export default function reducer(state, action) {
```
I'm surprised you don't get a warning about needing an initial state. I thought redux usually complained if calling `reducer(undefined, action)` returns undefined. I would make this `reducer(state=[], action) {`

pcardune replied:
I bet `_.cloneDeep(undefined)` returns an object.

joshlory replied:
Looks like `_.cloneDeep(undefined)` returns `undefined`.  I'm surprised Redux needs an initial state when I'm explicitly passing in the initial state when I create the store.

Bjvanminnen replied:
Oh, if you're passing in an initialstate when you create it, it probably doesn't. That would explain why you don't get warnings :)

Bjvanminnen commented on apps/src/sites/studio/pages/components/editorRedux.js:
```diff
@@ -0,0 +1,127 @@
+      newState.push({
+        id: newStageId--,
+        flex_category: prompt('Enter new group name'),
+        name: prompt('Enter new stage name'),
```
Having prompts in our reducer seems bad. We should prompt before getting here, and have the result of the prompts passed via our action.

pcardune commented on apps/src/sites/studio/pages/components/editorRedux.js:
```diff
@@ -0,0 +1,127 @@
+  let newState = _.cloneDeep(state);
+
+  switch (action.type) {
+    case 'REORDER_LEVEL': {
```
consider prefixing these action type strings with something indicating what domain these actions are relevant for. This makes it easier to combine multiple reducers in a larger application without worrying about action type collisions.

Bjvanminnen commented on apps/src/sites/studio/pages/components/editorRedux.js:
```diff
@@ -0,0 +1,127 @@
+
+function updatePositions(node) {
+  for (var i = 0; i < node.length; i++) {
+    node[i].position = i + 1;
```
Why do we need to have `node[i].position` if it can already be derived based on it's position in the array?

Bjvanminnen commented on apps/src/sites/studio/pages/components/editorRedux.js:
```diff
@@ -0,0 +1,127 @@
+      start = categories.indexOf(swappedGroupName);
+      count = categories.filter(c => c === swappedGroupName).length;
+      newState.splice(action.direction === 'up' ? start : start + count, 0, ...swap);
+      console.log(newState.map(s => s.flex_category));
```
console.log meant to be here?

Bjvanminnen commented on apps/src/sites/studio/pages/components/editorRedux.js:
```diff
@@ -0,0 +1,127 @@
+      const swappedGroupName = newState[action.direction === 'up' ? index - 1 : index].flex_category;
+      start = categories.indexOf(swappedGroupName);
+      count = categories.filter(c => c === swappedGroupName).length;
+      newState.splice(action.direction === 'up' ? start : start + count, 0, ...swap);
```
This is all a big block of dense code. Might be worth some comments explaining what's happening.

Bjvanminnen commented on dashboard/app/views/scripts/_form.html.haml:
```diff
@@ -1,42 +1,27 @@
+  :javascript
+    const scriptData = #{(@script ? @script.summarize : {}).to_json};
+    const i18nData = #{(@script ? @script.summarize_i18n : {}).to_json};
+    const levelKeyList = #{Level.key_list.to_json};
```
We might be able to do something like we did here https://github.com/code-dot-org/code-dot-org/pull/11322 and avoid sticking these on window.

Bjvanminnen commented:
Had a lot of small things, so wanted to summarize a couple of the more important pieces of feedback here:
(1) Your reducer is not a pure function, and needs to be.
(2) Mutating your entire state tree every time in your reducer is undesirable. Rethinking the shape of your state and/or using Immutable.js might be helpful here.
(3) Would be great to follow the pattern we've used elsewhere of making a `const` for each action type, and then exporting an action creator
(4) Let's avoid passing data from dashboard to apps by just sticking it on window. Passing in via a data attribute on the script tag might be a reasonable option here.

Let me know if you have questions about any of this :)

joshlory commented:
@pcardune I've moved the components to `lib/script-editor`.

joshlory commented:
@Bjvanminnen I believe this is ready for another pass.  I've addressed most of your feedback but I'm still working on adding Immutable.js / not cloning the entire state tree for each action.

Bjvanminnen commented on apps/src/lib/script-editor/FlexGroup.jsx:
```diff
@@ -0,0 +1,190 @@
+  serializeStages(stages) {
+    let s = [];
+    stages.forEach(stage => {
+      let t = `stage '${stage.name}'`;
```
could we have more descriptive variable names than s and t? (Here and in serializeLevel)

Bjvanminnen commented on apps/src/lib/script-editor/FlexGroup.jsx:
```diff
@@ -0,0 +1,190 @@
+    return s;
+  },
+
+  normalizeLevelKind(kind) {
```
Could be worth a comment describing why were doing this.

Bjvanminnen commented on apps/src/lib/script-editor/FlexGroup.jsx:
```diff
@@ -0,0 +1,190 @@
+    return (
+      <div>
+        {_.map(groups, (stages, group) => {
+          return (
```
nit, I think you could reduce indentation if you wanted by getting rid of the return statement and doing:

```
{_.map(groups, (stages, group) => (
  <div key={group}>
    ...
  </div>
)}
```

Bjvanminnen commented on apps/src/lib/script-editor/LevelTokenDetails.jsx:
```diff
@@ -0,0 +1,226 @@
+        {this.containsLegacyLevel() &&
+          <div>
+            <span
+              style={Object.assign(
```
Could use object destructuring instead of Object.assign (here and a few other places)
`style={{display: 'inline-block'}, ...styles.levelFieldLabel}}`

Bjvanminnen replied:
Actually, you repeat that same block a few times, and none of this is dynamic config. Shouldn't display: inline-block just be part of styles.levelFieldLabel?

Bjvanminnen commented on apps/src/lib/script-editor/LevelTokenDetails.jsx:
```diff
@@ -0,0 +1,226 @@
+                style={styles.checkbox}
+                name={`radio-${this.props.stagePosition}-${this.props.level.position}`}
+              />
+            </div>
```
I think we're a little bit inconsistent about this in our code (and styleguide might even say both are okay), but I think I've come around to agree with brad that it's better in cases like these to indent our JSX when part of a conditional continuation, i.e.

``` jsx
{this.props.somethingIsTrue &&
  <MyDiv/>
}
// instead of
{this.props.somethingIsTrue &&
<MyDiv/>
}
```

joshlory replied:
Good catch, I missed this indent!

Bjvanminnen commented on apps/src/lib/script-editor/StageCard.jsx:
```diff
@@ -0,0 +1,187 @@
+        </div>
+        {this.props.stage.levels.map(level =>
+          <LevelToken
+            ref={`levelToken${level.position}`}
```
In the first iteration, I think Paul made the comment that string refs are considered legacy (which I was somewhat dismissive of). I wonder tho if this is a case where we could actually benefit from non-string refs. I think the way to do this would be to do something like:

``` js
ref={levelToken => this.metrics[level.position] = levelToken.getClientBoundingRect()}
// then elsewhere you can do
this.metrics[token]; // instead of this.metrics(token)
```

(I don't feel that strongly about this, but it might be good for us to start getting away from using string refs).

joshlory replied:
Cool, this is much cleaner!

joshlory replied:
I wonder if this is abusing `ref` to create unexpected side effects?

Bjvanminnen replied:
I think this is the intention of ref (though I could be wrong). https://facebook.github.io/react/docs/refs-and-the-dom.html

Another option would be to behave more like in the link above, and just do `levelToken = this.levelTokens[position] = levelToken` and then in metrics do `this.levelTokens[position].getClientBoundingRect()`

Bjvanminnen commented on apps/src/lib/script-editor/editorRedux.js:
```diff
@@ -0,0 +1,235 @@
+export const NEW_LEVEL_ID = -1;
+
+export default function reducer(state, action) {
+  let newState = _.cloneDeep(state.stages);
```
Let me know if you run into any problems converting this to use Immutable.js

Bjvanminnen commented on apps/src/sites/studio/pages/levelbuilder_edit_script.js:
```diff
@@ -1,131 +1,47 @@
+  flex_category: stage.flex_category,
+  lockable: stage.lockable,
+  name: stage.name,
+  levels: stage.levels.filter(level => !level.uid || /_0$/.test(level.uid)).map(level => ({
```
What is this filter doing (and why)? Might deserve a comment.

joshlory commented:
Synced with @Bjvanminnen offline.  Planning to merge this without the Immutable.js changes, to be added in a follow up PR.
